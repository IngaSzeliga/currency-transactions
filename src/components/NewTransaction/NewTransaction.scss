.new-transaction-container {
  display: flex;
  flex-direction: column;
  padding: 40px;
  @media only screen and (min-width: 768px) {
    width: 80%;
  }

  .new-transaction-currency-rate {
    display: flex;
    align-items: baseline;
    .new-transaction-currency-rate-input {
      margin-left: 10px;
    }
  }
  .new-transaction-button-container {
    display: flex;
    justify-content: center;
    .new-transaction-button {
      margin-top: 50px;
      width: 150px;
    }
  }
}
