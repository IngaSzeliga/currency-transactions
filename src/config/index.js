export const countryRegion = "pl-PL";

export const E_KEYCODE = 69;
