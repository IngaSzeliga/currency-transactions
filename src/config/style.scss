$light-gray: #bdbdbd;
$light-black: #424242;
$dark-red: #d32f2f;
