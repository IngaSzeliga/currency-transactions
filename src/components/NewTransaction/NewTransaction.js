import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import uuidv4 from "uuid/v4";
import { Paper, TextField, Button } from "@material-ui/core";
import { E_KEYCODE, errorMessage } from "../../config";
import "./NewTransaction.scss";

class NewTransaction extends PureComponent {
  state = {
    transactionName: "",
    amount: "",
    rate: "",
    calculatedAmount: "",
    transactionNameError: "",
    amountError: "",
    rateError: ""
  };

  calculateAmount = (amount, rate) => amount * rate;

  handleTransactionName = event => {
    this.setState({ transactionName: event.target.value });
  };

  handleChangeRate = event => {
    const { amount } = this.state;
    const rate = Number(event.target.value);
    const calculatedAmount = this.calculateAmount(amount, rate);
    this.setState({ rate, calculatedAmount });
  };

  handleKeyDownRate = event => {
    if (event.keyCode === E_KEYCODE) {
      event.preventDefault();
    }
  };

  handleChangeAmount = event => {
    const { rate } = this.state;
    const amount = Number(event.target.value);
    const calculatedAmount = this.calculateAmount(amount, rate);
    this.setState({ amount, calculatedAmount });
  };

  handleSubmit = () => {
    const { transactionName, amount, calculatedAmount, rate } = this.state;
    const stateError = {
      transactionNameError: "",
      amountError: "",
      rateError: ""
    };
    let hasError = false;
    if (transactionName === "") {
      stateError.transactionNameError = errorMessage;
      hasError = true;
    }
    if (amount === "") {
      stateError.amountError = errorMessage;
      hasError = true;
    }
    if (rate === "") {
      stateError.rateError = errorMessage;
      hasError = true;
    }
    if (hasError) {
      this.setState(stateError);
    } else {
      const { addTransaction } = this.props;
      addTransaction({
        id: uuidv4(),
        title: transactionName,
        amount: amount,
        convertedAmount: calculatedAmount
      });
      this.setState({
        transactionName: "",
        amount: "",
        rate: "",
        calculatedAmount: ""
      });
    }
  };

  render() {
    const { transactionName, amount, rate, calculatedAmount } = this.state;
    const { transactionNameError, amountError, rateError } = this.state;
    return (
      <Paper className="new-transaction-container">
        <TextField
          label="Transaction Name"
          onChange={this.handleTransactionName}
          type="text"
          margin="normal"
          value={transactionName}
          error={transactionNameError !== ""}
          helperText={transactionNameError}
        />
        <div className="new-transaction-currency-rate">
          1EUR =
          <TextField
            className="new-transaction-currency-rate-input"
            onChange={this.handleChangeRate}
            onKeyDown={this.handleKeyDownRate}
            type="number"
            margin="normal"
            value={rate}
            InputProps={{ inputProps: { min: 0 } }}
            error={rateError !== ""}
            helperText={rateError}
          />
          PLN
        </div>
        <TextField
          label="Amount in Euro"
          onChange={this.handleChangeAmount}
          type="number"
          margin="normal"
          value={amount}
          error={amountError !== ""}
          helperText={amountError}
        />
        <div>Amount in PLN: {calculatedAmount}</div>
        <div className="new-transaction-button-container">
          <Button
            className="new-transaction-button"
            variant="contained"
            color="primary"
            onClick={this.handleSubmit}
          >
            Submit
          </Button>
        </div>
      </Paper>
    );
  }
}

NewTransaction.propTypes = {
  addTransaction: PropTypes.func.isRequired
};

export default NewTransaction;
