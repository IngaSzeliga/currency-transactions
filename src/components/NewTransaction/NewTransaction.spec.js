import React from "React";
import { shallow, mount } from "enzyme";
import NewTransaction from ".";
import { errorMessageInput } from "../../config";

describe("NewTransaction component", () => {
  const component = shallow(<NewTransaction addTransaction={jest.fn()} />);

  it("has working component", () => {
    expect(component).toBeDefined();
  });

  it("expects component to never be changed", () => {
    expect(component).toMatchSnapshot();
  });

  it("expects to call addTransaction", () => {
    const mockFunction = jest.fn();
    const componentMounted = mount(
      <NewTransaction addTransaction={mockFunction} />
    );
    componentMounted.setState({ transactionName: "Name", amount: 4, rate: 3 });
    componentMounted
      .find(".new-transaction-button")
      .first()
      .simulate("click");
    expect(mockFunction).toBeCalled();
  });

  it("expects to set an error because of the transactionName, amount and rate", () => {
    const componentMounted = mount(
      <NewTransaction addTransaction={jest.fn()} />
    );
    componentMounted.setState({ transactionName: "", amount: "", rate: "" });
    componentMounted
      .find(".new-transaction-button")
      .first()
      .simulate("click");
    expect(componentMounted.state().transactionNameError).toBe(
      errorMessageInput
    );
    expect(componentMounted.state().amountError).toBe(errorMessageInput);
    expect(componentMounted.state().rateError).toBe(errorMessageInput);
  });

  it("expects to fetch the rates", () => {
    const componentMounted = mount(
      <NewTransaction addTransaction={jest.fn()} />
    );
    componentMounted.setState({ amount: "" });
    componentMounted
      .find(".rate-button")
      .first()
      .simulate("click");

    setTimeout(() => {
      expect(componentMounted.state().rate).toBe(4.2843);
      expect(componentMounted.state().calculatedAmount).toBe("");
    }, 0);
  });

  it("expects to fetch the rates with amount", () => {
    const componentMounted = mount(
      <NewTransaction addTransaction={jest.fn()} />
    );
    componentMounted.setState({ amount: 33, rate: 12 });
    componentMounted
      .find(".rate-button")
      .first()
      .simulate("click");

    setTimeout(() => {
      expect(componentMounted.state().rate).toBe(4.2843);
      expect(componentMounted.state().calculatedAmount).toBe("396,00 z≈Ç");
    }, 0);
  });
});
