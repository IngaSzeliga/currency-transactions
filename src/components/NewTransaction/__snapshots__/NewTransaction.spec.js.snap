// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTransaction component expects component to never be changed 1`] = `
<div
  className="new-transaction-container"
>
  <WithStyles(ForwardRef(Paper))
    className="new-transaction-container-paper"
  >
    <WithStyles(ForwardRef(TextField))
      error={false}
      helperText=""
      label="Transaction Name"
      margin="normal"
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="new-transaction-currency-rate"
    >
      <div
        className="new-transaction-currency-rate-input-container"
      >
        1EUR =
        <WithStyles(ForwardRef(TextField))
          InputProps={
            Object {
              "inputProps": Object {
                "min": 0,
              },
            }
          }
          className="new-transaction-currency-rate-input"
          error={false}
          helperText=""
          margin="normal"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value=""
        />
        Â PLN
      </div>
      <WithStyles(ForwardRef(Button))
        className="rate-button"
        color="primary"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Get current rate
      </WithStyles(ForwardRef(Button))>
    </div>
    <WithStyles(ForwardRef(TextField))
      error={false}
      helperText=""
      label="Amount in Euro"
      margin="normal"
      onChange={[Function]}
      onKeyDown={[Function]}
      type="number"
      value=""
    />
    <div
      className="new-transaction-exchange-amount-container"
    >
      Amount in PLN:
    </div>
    <div
      className="new-transaction-button-container"
    >
      <WithStyles(ForwardRef(Button))
        className="new-transaction-button"
        color="primary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(ForwardRef(Button))>
    </div>
  </WithStyles(ForwardRef(Paper))>
</div>
`;
